### ISSUE TITLE
[Bug] Actions from outside action_space chosen

### ISSUE THREAD

**Issue by richleigh** [ray-project/ray#23074]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

RLlib

### Issue Severity

High: It blocks me to complete my task.

### What happened + What you expected to happen

This seems to be a regression on bug https://github.com/ray-project/ray/issues/2965

When given a box action space of -1 to +1, the agent is given actions outside of the [low, high] range.

Notably, config['clip_actions'] = True which as I understand it is the fix to the previously referenced issue.

### Versions / Dependencies

ray==1.11.0
ray-cpp==1.11.0 (from pypi)

Python 3.9.10 (cpython from homebrew)

MacOS Monterey 12.1 (x86_64, not arm)

### Reproduction script
```
import gym
import numpy as np
class BuggedEnv(gym.Env):
    def __init__(self, config):
        self.action_space = gym.spaces.Box(low=-1, high=1, shape=(100,), dtype=np.int32)
        self.observation_space = gym.spaces.Box(low=-10, high=10, shape=(40,), dtype=np.int32)
    def step(self, action):
        for idx, value in enumerate(action):
            assert -1 <= value <= 1, f"action {idx=} {value=} outside action space"
        return np.zeros(self.observation_space.shape, dtype=np.int32), 0, True, {}
    def reset(self):
        return np.zeros(self.observation_space.shape, dtype=np.int32)
import ray
ray.init()
import ray.rllib.agents.ppo as ppo
config = ppo.DEFAULT_CONFIG.copy()
config['clip_actions'] = True
agent = ppo.PPOTrainer(env=BuggedEnv, config=config)
agent.train() # Triggers assert on line 9
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by czgdp1807** [comment]
@pcmoritz @stephanie-wang Can I look into this?

**Comment by gjoliver** [comment]
@czgdp1807 appreciate your help

**Comment by czgdp1807** [comment]
Great. I will look into this tomorrow, Monday and Tuesday. Will keep this thread posted.

**Comment by czgdp1807** [comment]
I noticed that for the code in the OP `ray` internally uses `StochasticSampling` class which doesn't call `tf.clip_by_value` on `stochastic_actions` tensor. So, I tried doing the same (see diff below) after which the code in the OP worked fine. 

It is also worth noting that out of all samplers available, only `OrnsteinUhlenbeckNoise` calls `tf.clip_by_value` on `stochastic_actions`, rest of the samplers don't. 

```diff
diff --git a/rllib/utils/exploration/stochastic_sampling.py b/rllib/utils/exploration/stochastic_sampling.py
index c46116fcd..f8ab1fb2b 100644
--- a/rllib/utils/exploration/stochastic_sampling.py
+++ b/rllib/utils/exploration/stochastic_sampling.py
@@ -95,12 +95,16 @@ class StochasticSampling(Exploration):
             ),
             false_fn=lambda: action_dist.sample(),
         )
+
         deterministic_actions = action_dist.deterministic_sample()

         action = tf.cond(
             tf.constant(explore) if isinstance(explore, bool) else explore,
-            true_fn=lambda: stochastic_actions,
-            false_fn=lambda: deterministic_actions,
+            true_fn=lambda: tf.clip_by_value(stochastic_actions,
+                self.action_space.low * tf.ones_like(deterministic_actions),
+                self.action_space.high * tf.ones_like(deterministic_actions)
+            ),
+            false_fn=lambda: deterministic_actions
         )

         logp = tf.cond(
```

**Comment by czgdp1807** [comment]
Another thing to be noted here. The code in the OP goes through the following line. I noticed that `self.config` (which contains the `"clip_actions"` setting) isn't passed to `get_exploration_action` call below, so, either a sampler would always clip actions or it would never. I think we should send the config to sampler via `get_exploration_action` call so that it can decide whether to clip or not depending on the user requirements.

https://github.com/ray-project/ray/blob/9655851f32d68013734dd44321914b27ad8fe633/rllib/policy/dynamic_tf_policy.py#L366-L371

**Comment by fcampa7** [comment]
We were seeing actions outside of the range while using compute_actions and compute_single_action.

from ray.rllib.utils.test_utils import (
    check,
    check_compute_single_action,
    #check_train_results
)

import gym
import numpy as np
class BuggedEnv(gym.Env):
    def __init__(self, config):
        self.action_space = gym.spaces.Box(low=-1, high=1, shape=(100,), dtype=np.int32)
        self.observation_space = gym.spaces.Box(low=-10, high=10, shape=(40,), dtype=np.int32)
    def step(self, action):
        #for idx, value in enumerate(action):
            #assert -1 <= value <= 1, print("action {} {} outside action space".format(idx, value))
        return np.zeros(self.observation_space.shape, dtype=np.int32), 0, True, {}
    def reset(self):
        return np.zeros(self.observation_space.shape, dtype=np.int32)
import ray
ray.init()
import ray.rllib.agents.ppo as ppo
config = ppo.DEFAULT_CONFIG.copy()
#config['clip_actions'] = True
agent = ppo.PPOTrainer(env=BuggedEnv, config=config)
agent.train() # Triggers assert on line 9
check_compute_single_action(agent)

/anaconda/envs/drl/lib/python3.9/site-packages/ray/rllib/utils/test_utils.py", line 406, in _test
    raise ValueError(
ValueError: Returned action ([1 1 2 1 2 0 0 2 2 0 2 2 0 2 0 1 0 0 2 0 0 0 2 2 1 1 1 0 1 2 1 1 0 1 1 0 1
 1 1 1 2 1 2 1 1 1 0 2 1 2 0 2 1 1 0 2 2 2 0 2 0 2 1 0 1 1 0 1 0 2 0 0 2 0
 1 1 1 2 2 0 2 1 0 2 1 0 2 0 1 0 1 2 2 0 2 1 1 1 2 0]) of trainer/policy PPO not in Env's action_space Box(-1, 1, (100,), int32)

**Comment by fcampa7** [comment]
This is a more representative script of our problem specifically the action space.  

import ray
import gym
import numpy as np
import ray.rllib.agents.ppo as ppo
from ray.tune.logger import pretty_print
from ray.rllib.utils.test_utils import (
    check,
    check_compute_single_action,
    #check_train_results
)

from gym.spaces import Box, Discrete, Tuple, Dict

class Env(gym.Env):
    def __init__(self, config):
        #self.action_space = gym.spaces.Box(low=-1, high=1, shape=(100,), dtype=np.int32)
        #self.action_space = gym.spaces.Box(low=-1., high=1., shape=(1,), dtype=np.float32)
        #self.action_space = gym.spaces.Box(low=0, high=2, shape=(100,), dtype=np.int32)
        
        self.action_space = Dict({
                'a': Discrete(3),
                'b': Box(-1.,1., (1,)),
                'c': Box(-1.,1., (1,))
        })
        
        self.observation_space = gym.spaces.Box(low=-10., high=10., shape=(10,), dtype=np.float32)
    def step(self, action):
        b = action['b']
        if(b > 1. or b < -1.):
            print("b: {}".format(b))
        c = action['c']
        if(c > 1. or c < -1.):
            print("c: {}".format(c))
        #for idx, value in enumerate(action):
            #assert -1. <= value <= 1., print("action {} {} outside action space".format(idx, value))
            #assert 0 <= value <= 2, print("action {} {} outside action space".format(idx, value))
        return np.zeros(self.observation_space.shape, dtype=np.float32), 0, True, {}
    def reset(self):
        return np.zeros(self.observation_space.shape, dtype=np.float32)

ray.init()
config = ppo.DEFAULT_CONFIG.copy()
env = Env(config)
agent = ppo.PPOTrainer(env=Env, config=config)

for i in range(1):
  print('epoch: {}'.format(i))
  result = agent.train()
  print(pretty_print(result))

check_compute_single_action(agent)

obs = env.reset() 
done = False
for i in range(10):
  action = agent.compute_action(obs)
  #print(action)
  if action['b'] > 1. or action['b'] < -1.:
    print('b: {}'.format(action['b']))
  if action['c'] > 1. or action['c'] < -1.:
    print('c: {}'.format(action['c']))
  #action = agent.compute_single_action(obs)
  obs, reward, done, info = env.step(action)

Python 3.97
Ray 1.9.2 
Gym 0.18.3
Numpy 1.21.2
Tensorflow 2.4.1 
Matplotlib 3.5.0 
Pandas 1.3.5
Aiohttp 3.8.1
Dm-tree 0.1.6
Tabulate 0.8.9
Scikit-image 0.19.1
Lz4 3.1.10 

Without the stochastic sampling fix, I get these errors with ray 1.9.2. 
 /anaconda/envs/drl/lib/python3.9/site-packages/ray/rllib/utils/test_utils.py", line 406, in _test
    raise ValueError(
ValueError: Returned action ({'a': 2, 'b': array([1.0841328], dtype=float32), 'c': array([0.8843014], dtype=float32)}) of trainer/policy PPO not in Env's action_space Dict(a:Discrete(3), b:Box(-1.0, 1.0, (1,), float32), c:Box(-1.0, 1.0, (1,), float32))

With the stochastic sampling fix, I get these errors with ray 1.9.2. 
/anaconda/envs/drl/lib/python3.9/site-packages/ray/rllib/utils/exploration/stochastic_sampling.py", line 95, in <lambda>
(RolloutWorker pid=3576824)     self.action_space.low * tf.ones_like(deterministic_actions),
(RolloutWorker pid=3576824) AttributeError: 'Dict' object has no attribute 'low'

Python 3.97
Ray 1.11.0
Gym 0.18.3
Numpy 1.21.2
Tensorflow 2.4.1 
Matplotlib 3.5.0 
Pandas 1.3.5
Aiohttp 3.8.1
Dm-tree 0.1.6
Tabulate 0.8.9
Scikit-image 0.19.1
Lz4 3.1.10 

With stochastic sampling fix, I get these errors with ray 1.11.0.  

/anaconda/envs/drl_act/lib/python3.9/site-packages/ray/rllib/utils/test_utils.py", line 416, in _test
    raise ValueError(
ValueError: Returned action ({'a': 0, 'b': array([0.47641268], dtype=float32), 'c': array([-1.1895902], dtype=float32)}) of trainer/policy PPOTrainer not in Env's action_space Dict(a:Discrete(3), b:Box(-1.0, 1.0, (1,), float32), c:Box(-1.0, 1.0, (1,), float32))

Without stochastic sampling fix, I get these errors with ray 1.11.0.

/anaconda/envs/drl_act/lib/python3.9/site-packages/ray/rllib/utils/test_utils.py", line 416, in _test
    raise ValueError(
ValueError: Returned action ({'a': 1, 'b': array([0.7697576], dtype=float32), 'c': array([1.4896199], dtype=float32)}) of trainer/policy PPOTrainer not in Env's action_space Dict(a:Discrete(3), b:Box(-1.0, 1.0, (1,), float32), c:Box(-1.0, 1.0, (1,), float32))

**Comment by gjoliver** [comment]
I put up a fix https://github.com/ray-project/ray/pull/23941
